mam/src/mam.rs

```diff
@@ -39,116 +37,148 @@ where
+        payload.extend_from_slice(&key);
+        payload.extend_from_slice(&pascal::encode(siblings.len() / HASH_LENGTH));
+        payload.extend_from_slice(&siblings);
+        let end = payload.len();
+        mask_slice(&mut payload[cursor..end], sponge);
+        sponge.reset();
     }
-    (payload, root)
+    payload
+}
+enum UnmaskPascalStatus {
+    /// Message Hash did not have any hamming weight of zero
+    Searching,
+    /// Signature did not match expected root
+    Finalizing,
+    /// Array was too short
+    Done,
 }
 
 pub fn parse<C>(
     payload: &[Trit],
+    side_key: &[Trit],
     root: &[Trit],
     index: usize,
-    curl1: &mut C,
-    curl2: &mut C,
+    sponge: &mut C,
 ) -> Result<(Vec<Trit>, Vec<Trit>), MamError>
 where
     C: Curl<Trit>,
 {
-    let mut channel_key: [Trit; HASH_LENGTH] = [0; HASH_LENGTH];
-    channel_key.clone_from_slice(&root[..HASH_LENGTH]);
-    add_assign(&mut channel_key, index as isize);
-    let mut unmasked_payload = payload.to_vec();
-    unmask::<C>(&mut unmasked_payload, &channel_key, curl1);
-
-    curl1.reset();
-    authenticate::<C>(&unmasked_payload, root, index, curl1, curl2)
+    let mut hash: [Trit; HASH_LENGTH] = [0; HASH_LENGTH];
+    sponge.reset();
+    message_key(side_key, root, index, sponge);
+
+    let (message_length, message_length_end) = pascal::decode(&payload);
+    if message_length > payload.len() {
+        return Err(MamError::ArrayOutOfBounds);
+    }
+    sponge.absorb(&payload[..message_length_end]);
+    let mut out: Vec<Trit> = (&payload[message_length_end..payload.len()])
+        .clone()
+        .to_vec();
+    let mut pos = {
+        unmask_slice(&mut out[..message_length], sponge);
+        message_length
+    };
+    unmask_slice(&mut out[pos..pos + HASH_LENGTH / 3], sponge);
+    pos += HASH_LENGTH / 3;
+    let mut hmac: [Trit; HASH_LENGTH] = [0; HASH_LENGTH];
+    hmac.clone_from_slice(&sponge.rate());
+    let security = iss::checksum_security(&hmac);
+    let out_len = out.len();
+    unmask_slice(&mut out[pos..out_len], sponge);
+    hash.clone_from_slice(sponge.rate());
+    if security != 0 {
+        let sig_end = pos + security * iss::KEY_LENGTH;
+        iss::digest_bundle_signature(&hmac, &mut out[pos..sig_end], sponge);
+        hash.clone_from_slice(&sponge.rate());
+        sponge.reset();
+        iss::address(&mut hash, sponge);
+        pos = sig_end;
+        let l = pascal::decode(&out[pos..]);
+        pos += l.1;
+        let sib_end = pos + l.0 * HASH_LENGTH;
+        let siblings = &out[pos..sib_end];
+        let end = merkle::root(&hash, siblings, index, sponge);
+        hash.clone_from_slice(&sponge.rate());
+
+        if hash.iter()
+            .zip(root.iter())
+            .take_while(|&(&a, &b)| a != b)
+            .count() == 0
+        {
+            let next_root: Vec<Trit> = out[..HASH_LENGTH].to_vec();
+            let message_out: Vec<Trit> = out[HASH_LENGTH..message_length].to_vec();
```

Comment:
This will cause unneccessary reallocation.
The most resource conservative solution is probably:
```rust
let message_out: Vec<Trit> = out[HASH_LENGTH..message_length].to_vec();
let next_root: Vec<Trit> = out.truncate(HASH_LENGTH);
```